crates/aptos-jwk-consensus/src/jwk_manager/mod.rs

```diff
@@ -0,0 +1,297 @@
+    validator_txn::ValidatorTransaction,
+};
+use aptos_validator_transaction_pool as vtxn_pool;
+use certified_update_producer::CertifiedUpdateProducer;
+use futures_util::future::AbortHandle;
+use std::{
+    collections::{BTreeMap, HashMap},
+    sync::Arc,
+};
+
+pub mod certified_update_producer;
+
+/// `JWKManager` executes per-issuer JWK consensus sessions
+/// and updates validator txn pool with quorum-certified JWK updates.
+pub struct JWKManager {
+    signing_key: PrivateKey,
+    my_addr: AccountAddress,
+    epoch_state: EpochState,
+    certified_update_producer: Arc<dyn CertifiedUpdateProducer>,
+    certified_update_tx: Option<aptos_channel::Sender<(), QuorumCertifiedUpdate>>,
+    vtxn_pool_write_cli: Arc<vtxn_pool::SingleTopicWriteClient>,
+    states_by_issuer: HashMap<Issuer, PerProviderState>,
+    _stopped: bool,
+}
+
+impl JWKManager {
+    pub fn new(
+        signing_key: PrivateKey,
+        my_addr: AccountAddress,
+        epoch_state: EpochState,
+        certified_update_producer: Arc<dyn CertifiedUpdateProducer>,
+        vtxn_pool_write_cli: Arc<vtxn_pool::SingleTopicWriteClient>,
+    ) -> Self {
+        Self {
+            signing_key,
+            my_addr,
+            epoch_state,
+            certified_update_producer,
+            certified_update_tx: None,
+            vtxn_pool_write_cli,
+            states_by_issuer: HashMap::default(),
+            _stopped: false,
+        }
+    }
+
+    /// Triggered by an observation thread periodically.
+    pub fn process_new_observation(&mut self, issuer: Issuer, jwks: Vec<JWK>) -> Result<()> {
+        let state = self.states_by_issuer.entry(issuer.clone()).or_default();
+        state.observed = Some(jwks.clone());
+        if state.observed.as_ref() != state.on_chain.as_ref().map(ProviderJWKs::jwks) {
+            let observed = ProviderJWKs {
+                issuer: issuer.clone(),
+                version: state.on_chain_version() + 1,
+                jwks: jwks.clone(),
+            };
+            let signature = self
+                .signing_key
+                .sign(&observed)
+                .map_err(|e| anyhow!("crypto material error occurred duing signing: {}", e))?;
+            let abort_handle = self.certified_update_producer.start_produce(
+                self.epoch_state.clone(),
+                observed.clone(),
+                self.certified_update_tx.clone(),
+            );
+            state.consensus_state = ConsensusState::InProgress {
+                my_proposal: ObservedUpdate {
+                    author: self.my_addr,
+                    observed,
+                    signature,
+                },
+                abort_handle_wrapper: AbortHandleWrapper::new(abort_handle),
+            };
+        }
+
+        Ok(())
+    }
+
+    /// Invoked on start, or on on-chain JWK updated event.
+    pub fn reset_with_on_chain_state(&mut self, on_chain_state: ObservedJWKs) -> Result<()> {
```

Comment:
shall we do per issuer reset?

Reply:
will add a TODO. We can come back to this after we land smoke tests.